src/sources/exec/mod.rs

```diff
@@ -0,0 +1,798 @@
+    }))
+}
+
+async fn run_command(
+    config: ExecConfig,
+    hostname: Option<String>,
+    shutdown: ShutdownSignal,
+    mut out: Pipeline,
+) -> Result<Option<ExitStatus>, Error> {
+    debug!("Starting command run.");
+    let mut command = build_command(&config);
+
+    // Mark the start time just before spawning the process as
+    // this seems to be the best approximation of exec duration
+    let start = Instant::now();
+
+    let mut child = command.spawn()?;
+
+    // Set up communication channels
+    let (sender, mut receiver) = channel(1024);
+
+    // Optionally include stderr
+    if config.include_stderr {
+        let stderr = child.stderr.take().ok_or_else(|| {
+            Error::new(ErrorKind::Other, "Unable to take stderr of spawned process")
+        })?;
+
+        // Create stderr async reader
+        let stderr = stderr.allow_read_until(shutdown.clone().map(|_| ()));
+        let stderr_reader = BufReader::new(stderr);
+
+        spawn_reader_thread(
+            stderr_reader,
+            config.event_per_line,
+            config.maximum_buffer_size_bytes,
+            STDERR,
+            sender.clone(),
+        );
+    }
+
+    let stdout = child
+        .stdout
+        .take()
+        .ok_or_else(|| Error::new(ErrorKind::Other, "Unable to take stdout of spawned process"))?;
+
+    // Create stdout async reader
+    let stdout = stdout.allow_read_until(shutdown.clone().map(|_| ()));
+    let stdout_reader = BufReader::new(stdout);
+
+    let pid = child.id();
+
+    spawn_reader_thread(
+        stdout_reader,
+        config.event_per_line,
+        config.maximum_buffer_size_bytes,
+        STDOUT,
+        sender,
+    );
+
+    while let Some((line, stream)) = receiver.recv().await {
+        let event = create_event(&config, &hostname, line, &Some(stream.to_string()), pid);
+
+        let _ = out
+            .send(event)
+            .await
+            .map_err(|_: crate::pipeline::ClosedError| {
+                error!(message = "Failed to forward events; downstream is closed.");
+            });
+    }
+
+    let elapsed = start.elapsed();
+
+    let result = match child.try_wait() {
+        Ok(Some(exit_status)) => {
+            handle_exit_status(&config, Some(exit_status), elapsed).await;
+            Ok(Some(exit_status))
+        }
+        Ok(None) => {
+            handle_exit_status(&config, None, elapsed).await;
+            Ok(None)
+        }
+        Err(error) => {
+            error!(message = "Unable to obtain exit status.", %error);
+
+            handle_exit_status(&config, None, elapsed).await;
+            Ok(None)
+        }
+    };
+
+    debug!("Finished command run.");
+    let _ = out.flush().await;
+
+    result
+}
+
+async fn handle_exit_status(
+    config: &ExecConfig,
+    exit_status: Option<ExitStatus>,
+    exec_duration: Duration,
+) {
+    let exit_status = match exit_status {
+        Some(exit_status) => exit_status.code(),
+        None => None,
+    };
```

Comment:
This may be simpler to have the single caller that passes `Some` handle this above, which would eliminate this match block.
`handle_exit_status(&config, Some(exit_status.code()), elapsed)`

Reply:
Converted as you suggested. The `handle_exit_status` method has become simpler and simpler over the reviews but I thought it was still useful to keep it as a placeholder for future logic if needed e.g. if we ever decide to add back in an event with the exit status and duration.